lib/translator/read.cc

```diff
@@ -0,0 +1,232 @@
+}
+
+// Translates fields common to all Read commands
+// Named Legacy because it is called directly by Read6, an obsolete command
+// lacking fields common to other Read commands
+StatusCode LegacyRead(nvme::GenericQueueEntryCmd& nvme_cmd,
+                      Allocation& allocation, uint32_t nsid,
+                      uint32_t transfer_length_pages) {
+  StatusCode status = allocation.SetPages(transfer_length_pages, 0);
+  if (status != StatusCode::kSuccess) {
+    return status;
+  }
+
+  nvme_cmd = nvme::GenericQueueEntryCmd{
+      .opc = static_cast<uint8_t>(nvme::NvmOpcode::kRead),
+      .psdt = 0,  // PRPs are used for data transfer
+      .nsid = nsid};
+
+  nvme_cmd.dptr.prp.prp1 = allocation.data_addr;
+
+  return StatusCode::kSuccess;
+}
+
+// Translates fields common to Read10, Read12, Read16
+StatusCode Read(uint8_t rd_protect, bool fua, uint32_t transfer_length,
+                nvme::GenericQueueEntryCmd& nvme_cmd, Allocation& allocation,
+                uint32_t nsid, uint32_t page_size, uint32_t lba_size) {
+  if (transfer_length == 0) {
+    DebugLog("NVMe read command does not support transfering zero blocks");
+    return StatusCode::kNoTranslation;
+  }
+
+  // Since NVMe has a protocol limit of 16 bits on transfer size,
+  // we can enforce a transfer limit of 16 bits on our SCSI device
+  if (transfer_length > 0xffff) {
+    DebugLog("Transfer length exceeds limit of 16 bits");
+    return StatusCode::kInvalidInput;
+  }
+  transfer_length &= 0xffff;  // truncate to 16 bits
+
+  StatusCode status =
+      LegacyRead(nvme_cmd, allocation, nsid,
+                 GetTransferLengthPages(transfer_length, page_size, lba_size));
+  if (status != StatusCode::kSuccess) {
+    return status;
+  }
+
+  uint8_t prinfo;  // Protection Information field 4 bits;
+  status = BuildPrinfo(rd_protect, prinfo);
+  if (status != StatusCode::kSuccess) {
+    return status;
+  }
+
+  nvme_cmd.cdw[2] = htoll(BuildCdw12(transfer_length, prinfo, fua));
+
+  return StatusCode::kSuccess;
+}
+
+}  // namespace
+
+StatusCode Read6ToNvme(Span<const uint8_t> scsi_cmd,
+                       nvme::GenericQueueEntryCmd& nvme_cmd,
+                       Allocation& allocation, uint32_t nsid,
+                       uint32_t page_size, uint32_t lba_size) {
+  scsi::Read6Command read_cmd;
+  if (!ReadValue(scsi_cmd, read_cmd)) {
+    DebugLog("Malformed Read6 command");
+    return StatusCode::kInvalidInput;
+  }
+  // Transfer Length set to 0 specifies 256 logical blocks to be read
+  // Section 3.15 Seagate SCSI specs
+  uint16_t updated_transfer_length =
+      read_cmd.transfer_length == 0 ? 256 : read_cmd.transfer_length;
+
+  StatusCode status = LegacyRead(
+      nvme_cmd, allocation, nsid,
+      GetTransferLengthPages(updated_transfer_length, page_size, lba_size));
+  if (status != StatusCode::kSuccess) {
+    return status;
+  }
+  // cdw10 Starting lba bits 31:00
+  uint32_t host_endian_lba = (read_cmd.logical_block_address_1 << 16) +
```

Comment:
You're byte swapping lba_2 but not lba_1

Reply:
Never mind - it's because lba_1 is only one byte.  I keep forgetting read6 has a completely ridiculous 24 bit lba field.